### PR TITLE
sleeping carp now deflects projectiles in random directions

### PR DESCRIPTION
**Pull request by tralezab**
:cl:
tweak: sleeping carp now deflects projectiles
/:cl:

I got the idea from the "epic idea for sleeping carp changes" thread and yeah it turned out great

upsides: slight chance the assailant hits themselves with their own projectile
downsides(?): fireballs, explosions, everything like that no longer just deletes, it'll explode near you

might need to change the deflect sound though, it sounds like they are dodging the bullet yet it changes direction and all that now

### CODE REVIEW

**Comment by tralezab** [comment]
also to anybody who is wondering why I'm not using the code for juggernauts deflection, they really are more reflecting as it tries to send the projectile back towards the person using it instead of a random angle. a LOT more powerful.

**Comment by Shadowflame909** [comment]
Since this is really a nerf, and not a buff or anything. Because the old one just deleted the item, and now you can run into it/if near the radius of the attack blow yourself up.

I think you should lower the  cost amount of this item IMO

**Comment by Therapist1** [comment]
Wouldn’t that just deflect it again?

**Comment by tralezab** [comment]
Therapist1 is right, and also it sets you to the owner of the projectile so you can't even run into it anyways. when i was testing i lined up like 20 emitters and just ran through the bullets so i could do some kind of bullet moses thing (time not wasted, it was dope) and not once was i hit by "my" own projectile

**Comment by Shadowflame909** [comment]
Yeah, but I also remember hearing about that Sleeping Carp doesnt have a chance to 100 deflect all weapon types.

Maybe it was bullets?


And if there is no such thing, then I still point towards that sleeping carp users are now vulnerable to thrown bombs, when they would delete them previously.

**Comment by tralezab** [comment]
it do deflect all projectiles 100% chance

**Comment by Shadowflame909** [comment]
Well, I guess. I just want to see if this negatively affects any sleeping carp users first. I'm glad that you tested it.

**Comment by wesoda25** [comment]
It shouldn’t be random directions!! It should deflect but not reverse the path of the projectile!!

**Comment by necromanceranne** [comment]
Let's not think about how exactly you're slapping lasers back at people.

Also, the situation where carp users are fucked by explosions is either going to be some miner got the wand of fireball and didn't aim near the carp user rather than straight at him, frag-12 slugs and mech missiles (assuming they're projectiles and I think they are). That's probably about it.

**Comment by Shadowflame909** [comment]
IED's. Combustable Lemons, Bomb Spears

**Comment by tralezab** [comment]
i just think it looks really neat (and it do!), i didn't want it to become a balance thing

**Comment by imsxz** [comment]
@Shadowflame909 those aren't projectiles. the tl;dr of this is that explosive projectiles are few and far between, and it'll be really funny if a carpie deflects one next to himself and dies.

**Comment by wesoda25** [comment]
Gib video of you shooting at someone with carp.

**Comment by 81Denton** [comment]
People already use explosive lances against carp/shielded hardsuits/wizards/whatnot. This PR doesn't change that

**Comment by necromanceranne** [comment]
Now that i think about it, frag-12's have a really low explosion radius (so small that even firing into the wall adjacent to people does not necessarily mean they'll take damage) so frankly it's more likely for the carp user to slap it into someone else and the poor sap gets fucking obliterated while they're unscathed.

**Comment by Shadowflame909** [comment]
Alright, no further questions. I've been won over

**Comment by obscolene** [comment]
Next to nerf explosive lances.

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 1]
```diff
@@ -52,7 +52,12 @@
 					else
 						visible_message("<span class='danger'>[src] deflects the projectile!</span>", "<span class='userdanger'>You deflect the projectile!</span>")
 					playsound(src, pick('sound/weapons/bulletflyby.ogg', 'sound/weapons/bulletflyby2.ogg', 'sound/weapons/bulletflyby3.ogg'), 75, 1)
-					return 0
+					if(!mind.martial_art.reroute_deflection)
+						return 0
+					else
+						P.firer = src
+						P.setAngle(rand(1, 360))//SHING
```
Comment (ShizCalev) [inline code/modules/mob/living/carbon/human/human_defense.dm]: wouldn't 0 be valid?

Comment (tralezab) [reply code/modules/mob/living/carbon/human/human_defense.dm]: fixed

**Comment by optimumtact** [comment]
this is a real dumb martial art

**Comment by tralezab** [comment]
combos need a rework, they are all very boring and it does not matter which combo you do since they lead into more stuns. they are also not very interesting, at least one of the martial arts has cool animations.

**Comment by ExcessiveUseOfCobblestone** [comment]
> realistically if the sleeping carp artist is getting shot at, they are ready to use their martial arts and thus would have the better reflect at all times, so i'm just gonna go with this for now

Not true, they most likely have a device to stun the user (baton) if it's known they have carp.

Otherwise they'll just grab/punch you into death.

**Comment by tralezab** [comment]
https://www.youtube.com/watch?v=BNgwRw2O60c

I warned you, it's pretty fucking funny

**Comment by JohnMcLaime** [comment]
Why bother with combos when you can pull out a stunbaton and one hit stun - kill people. Combos do need the rework.

**Comment by wesoda25** [comment]
If anything thats an issue with stun weaponry. 

Why isnt this merged ahhhhhh

**Comment by ExcessiveUseOfCobblestone** [comment]
For me it's a bit difficult to merge this considering it was balanced behind not using ranged weapons. Being able to deflect range weapons back to the user, even if it's not a very high chance and assuming it doesn't hit anyone else, seems like it goes against that.

**Comment by tralezab** [comment]
I made it cuz it's cute and funny to see pretty things fly around but I acknowledge I didnt think about balance at all si I won't be hurt if you close it

**Comment by ExcessiveUseOfCobblestone** [comment]
We can try it out! If it's icky i'll just revert it.

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 2]
```diff
@@ -52,7 +52,12 @@
 					else
 						visible_message("<span class='danger'>[src] deflects the projectile!</span>", "<span class='userdanger'>You deflect the projectile!</span>")
 					playsound(src, pick('sound/weapons/bulletflyby.ogg', 'sound/weapons/bulletflyby2.ogg', 'sound/weapons/bulletflyby3.ogg'), 75, 1)
-					return 0
+					if(!mind.martial_art.reroute_deflection)
+						return 0
+					else
+						P.firer = src
+						P.setAngle(rand(0, 360))//SHING
+						return BULLET_ACT_FORCE_PIERCE
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/human/human_defense.dm]: Why does this proc accept return values of both strings and integers >:(

Comment (tralezab) [reply code/modules/mob/living/carbon/human/human_defense.dm]: i swear i saw return 0 somewhere else in the code but i suppose i was mistaken, i can fix that in a little bit

Comment (ExcessiveUseOfCobblestone) [reply code/modules/mob/living/carbon/human/human_defense.dm]: oh it isn't suppose to accept numbers, how funny.

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 3]
```diff
@@ -52,7 +52,12 @@
 					else
 						visible_message("<span class='danger'>[src] deflects the projectile!</span>", "<span class='userdanger'>You deflect the projectile!</span>")
 					playsound(src, pick('sound/weapons/bulletflyby.ogg', 'sound/weapons/bulletflyby2.ogg', 'sound/weapons/bulletflyby3.ogg'), 75, 1)
-					return 0
+					if(!mind.martial_art.reroute_deflection)
+						return 0
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/human/human_defense.dm]:
```suggestion
						return BULLET_ACT_BLOCK
```

**Pull request merged**